### ISSUE TITLE
AWS SCIM Provisioning Failure

### ISSUE THREAD

**Issue by arcdigital** [goauthentik/authentik#12321]
**Describe the bug**
AWS is returning 400s during the SCIM sync. It looks like authentik is trying to do a PATCH request that AWS doesn't like.

{"domain_url": null, "event": "Failed to send SCIM request", "level": "warning", "logger": "authentik.lib.sync.outgoing.base", "method": "PATCH", "path": "/Groups/__uuid__", "provider": "AWS SCIM", "response": "{\"schema\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"schemas\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"detail\":\"Current PATCH support allows omitting \\\"path\\\" attribute only for Add and Replace operations modifying a single attribute.\",\"status\":\"400\",\"exceptionRequestId\":\"__uuid__\",\"timeStamp\":\"2024-12-10 19:50:00.531\"}", "schema_name": "public", "task_id": "task-ID", "timestamp": "2024-12-10T19:50:00.539728"}

**Version and Deployment (please complete the following information):**
-   authentik version: 2024.10.4

**Comment by uptownhr** [comment]
I am currently using `2024.8.4` and also experiencing this issue.

**Comment by fullykubed** [comment]
@BeryJu It appears the problem are is [here](https://github.com/goauthentik/authentik/blame/dc3559c7e913d20ef656c5869464956751279edc/authentik/providers/scim/clients/groups.py#L123) which you last updated to introduce PATCH support.

The way that this is implemented sends a PATCH request with the full group model dumped as JSON. However, AWS only supports PATCH with a _single_ group property at a time.

As far as I can tell, this doesn't seem _too_ bad of a bug atm, but definitely not ideal. While groups can still be created, deleted, and have their membership change, you would not be able to update things like the name have then have that reflected in AWS.

Do you have any thoughts on whether a workaround is possible?

**Comment by fullykubed** [comment]
It would be kindof nice to be able to override the configuration values returned by the providers `/ServiceProviderConfig` endpoint to be able to deal with cases like this where the SCIM target has weird limitations ([code](https://docs.aws.amazon.com/singlesignon/latest/developerguide/serviceproviderconfig.html)).

To be fair to Authentik, it looks like AWS is the one that is the one that is not comformant with the [SCIM PATCH spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.5.2) currently.

**Comment by fullykubed** [comment]
It might make sense just to do a fallback to PUT if the PATCH doesn't work?

**Comment by fullykubed** [comment]
@BeryJu Would you be open to a PR that implements the retry / fallback to PUT if PATCH doesn't work?

**Comment by argo1984** [comment]
Are there any news regarding to it? Is there an existing workaround?

**Comment by jorhett** [comment]
> As far as I can tell, this doesn't seem _too_ bad of a bug atm, but definitely not ideal. While groups can still be created, deleted, and have their membership change, you would not be able to update things like the name have then have that reflected in AWS.

I have recently setup AWS SCIM and NO groups are successfully being created due to this bug, so yes it is TOO bad of a bug. It makes SCIM effectively worthless for AWS.

**Comment by fullykubed** [comment]
@jorhett Just for clarity, we regularly set up new authentik<->aws integrations and the initial provisioning seems to work fine. Not saying you aren't valid in your concern here (@BeryJu still happy to tackle this if you sign-off), but might be worth double checking something else isn't amiss in your setup.

**Comment by argo1984** [comment]
@fullykubed the initial sync seemed to work for me as well but i had some issues and resettet the syn. the "2nd initial sync" didn't deploy the groups. I resetted everything on aws side.

**Comment by fullykubed** [comment]
That makes sense. Good to know!

**Comment by jorhett** [comment]
@fullykubed I'm 100% willing to investigate anything it could be on our side. The reason I opened this issue is that I can't find anything left I can debug:

* AWS Is masking all the details so I don't even know what entry they say is a duplicate
* I can't seem to find any way to get higher levels of debug out of Authentik

I'm a coder happy and willing to identify the problem and even provide a patch if I can find a way to get more debug output to tell me what's going on here.

Looking at the cloudtrail logs it seems we are issuing more CreateUser requests when the user list hasn't change. So it seems likely that something isn't syncing up correctly, and authentik believes it needs to create the users.

Also, as @argo1984 experienced, our Groups are synced but have zero members.

**Comment by jorhett** [comment]
Just being able to get a cleartext copy of the SCIM transactions would really help.

**Comment by fullykubed** [comment]
@jorhett To be clear, I am just another user like yourself who has stumbled upon this unfortunate problem. I have done the necessary inspection of the network traffic to highlight the root cause [here](https://github.com/goauthentik/authentik/issues/12321#issuecomment-2558539103). I am also willing to allocate resources from my company to fix it. However, ultimately my hands are tied because Authentik's CTO @arcdigital is the one who introduced the code change and he hasn't been able / willing to respond to our offer to help assist in fixing this integration.

**Comment by arcdigital** [comment]
@fullykubed What? I'm neither Authentik's CTO, nor the person who introduced the code change??

**Comment by fullykubed** [comment]
Whoops! Meant @BeryJu ! See the above thread.

**Comment by kuzaxak** [comment]
I can confirm that monkey patching Docker image file `/authentik/providers/scim/clients/groups.py` to the version from commit eafb7093c946276cbca2418cdfd18e932836b249 fixing the issue with AWS. Without it, we won't be able to provision users to groups.

**Comment by argo1984** [comment]
HI @kuzaxak, it seems this patch is in the current version 2025.2.1 included, but the issue still persists. Do i need to change settings in the SCIM Provider?

**Comment by jorhett** [comment]
Yeah, I see the code change here to fall back https://github.com/goauthentik/authentik/blob/version-2025.2/authentik/providers/scim/clients/groups.py#L116-L118 but it's not working in practice.

```
Failed to sync Group Group example due to transient error: <Response [400]>	SCIM Provider@authentik.core.models.Group

Attributes: {
    "obj": {
        "pk": "012345",
        "app": "authentik_core",
        "name": "example",
        "model_name": "group"
    }
}
```

**Comment by jorhett** [comment]
For others who stumble onto this and wonder what _monkey patch_ means, here's how I used @kuzaxak's finding to get AWS groups syncing again: 

```shell
$ wget https://raw.githubusercontent.com/goauthentik/authentik/eafb7093c946276cbca2418cdfd18e932836b249/authentik/providers/scim/clients/groups.py
$ podman cp groups.py server:/authentik/providers/scim/clients/
$ podman cp groups.py worker1:/authentik/providers/scim/clients/
...
```

There's probably a smarter way to tweak the underlying image but this works for a quick monkey patch.

**Comment by jorhett** [comment]
Looking in more depth to the code base, actually nothing changed since Oct (sorry assumed it did based on the comment). It seems the code contains [an if/else for put/patch](https://github.com/goauthentik/authentik/blob/version-2025.2/authentik/providers/scim/clients/groups.py#L116-L118) but it appears that we get this value by [asking the provider](https://github.com/goauthentik/authentik/blob/8244c2340a88e31f8b242bff5d8a9774993fff42/authentik/providers/scim/clients/base.py#L89-L95)

I've confirmed by manual test that AWS returns `"patch":{"supported":true},` but as this issue explains it's incomplete support.

Perhaps the fix here is simply to add a UI element which allows the user to override this value as needed? Or just to hardcode that AWS doesn't play ball correctly?

**Comment by BeryJu** [comment]
a better/less hacky way to deal with this would be to add an option to override the data returned by `ServiceProviderConfig`, I'm really not a fan of having the hardcoded AWS hostnames in there

**Comment by arcdigital** [comment]
I agree, that patch doesn't include the domain for other AWS partitions, like GovCloud.

**Comment by jorhett** [comment]
> a better/less hacky way to deal with this would be to add an option to override the data returned by `ServiceProviderConfig`, 

you're passing it immediately into a method to verify that I couldn't track down in the code base so I dealt with it where I could.

> I'm really not a fan of having the hardcoded AWS hostnames in there

I labeled it a hack 😁  The problem here is that a major integration partner has been broken for 5 months. You're asking paying customers to go back to manual group manipulation until a perfect solution falls out of the sky somehow, which seems more than a bit unreasonable. A well-documented hack that works is better than forcing your customers to take their business elsewhere.

> I agree, that patch doesn't include the domain for other AWS partitions, like GovCloud.

I directly referenced AWS docs for where SCIM was supported. But if you know better, fixing it would take less characters than you spent complaining about it.

**Comment by jorhett** [comment]
> I'm really not a fan of having the hardcoded AWS hostnames in there

I'm really a fan, I mean a huge fan, of working code.  I'm also a fan of not having to work around my vendors who show disinterest in working code. (look how long this has been broken!) 

Consider this more of a prompt to do it better. I don't care how you code it, but continued inaction on this is not a workable situation.

**Comment by xabinapal** [comment]
I'm not sure if this has been fixed. Just updated to release `2025.4.0` and switched my SCIM provider to use the new **Compatibility mode** set to *AWS*. After triggering a resync via the UI, I'm still seeing multiple `400` responses from AWS.

![Image](https://github.com/user-attachments/assets/cf12412e-4141-43cd-a980-d54f6ec26529)

The error message in the logs is the same as before:

```
{"domain_url": null, "event": "Failed to send SCIM request", "level": "warning", "logger": "authentik.lib.sync.outgoing.base", "method": "PATCH", "path": "/Groups/[REDACTED]", "pid": 478, "provider": "AWS SCIM", "response": "{\"schema\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"schemas\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"detail\":\"Current PATCH support allows omitting \\\"path\\\" attribute only for Add and Replace operations modifying a single attribute.\",\"status\":\"400\",\"exceptionRequestId\":\"[REDACTED]\",\"timeStamp\":\"2025-05-08 16:09:34.099\"}", "schema_name": "public", "task_id": "task-[REDACTED]", "timestamp": "2025-05-08T16:09:34.155397"}

{"domain_url": null, "event": "failed to sync object", "exc": "SCIMRequestException(<Response [400]>)", "level": "warning", "logger": "authentik.lib.sync.outgoing.tasks", "object_type": "authentik.core.models.Group", "pid": 478, "provider_pk": 19, "provider_type": "authentik.providers.scim.models.SCIMProvider", "schema_name": "public", "task_id": "task-[REDACTED]", "timestamp": "2025-05-08T16:09:34.155618", "user": "<Group: Group [REDACTED]>"}

{"domain_url": null, "event": "Failed to send SCIM request", "level": "warning", "logger": "authentik.lib.sync.outgoing.base", "method": "PUT", "path": "/Groups/[REDACTED]", "pid": 478, "provider": "AWS SCIM", "response": "", "schema_name": "public", "task_id": "task-[REDACTED]", "timestamp": "2025-05-08T16:09:34.045554"}
```

**Comment by FilipK-CZ** [comment]
I still have same issue on 2025.6.2

```
{"domain_url": null, "event": "Failed to send SCIM request", "level": "warning", "logger": "authentik.lib.sync.outgoing.base", "method": "PATCH", "path": "/Groups/7275f414-0000-0000-0000-4b237bcd7445", "pid": 333, "provider": "AWS SCIM", "response": "{\"schema\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"schemas\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"detail\":\"Current PATCH support allows omitting \\\"path\\\" attribute only for Add and Replace operations modifying a single attribute.\",\"status\":\"400\",\"exceptionRequestId\":\"bd75928a-0000-0000-0000-d525fa850aec\",\"timeStamp\":\"2025-06-17 10:21:32.16\"}", "schema_name": "public", "task_id": "task-213e517f1eb0000000000856fddf59e6", "timestamp": "2025-06-17T10:21:32.161572"}
{"domain_url": null, "event": "failed to sync object", "exc": "SCIMRequestException(<Response [400]>)", "level": "warning", "logger": "authentik.lib.sync.outgoing.tasks", "object_type": "authentik.core.models.Group", "pid": 333, "provider_pk": 24, "provider_type": "authentik.providers.scim.models.SCIMProvider", "schema_name": "public", "task_id": "task-213e517f1eb0000000000856fddf59e6", "timestamp": "2025-06-17T10:21:32.161830", "user": "<Group: Group XXXXXXX>"}
```

**Comment by FilipK-CZ** [comment]
AWS CloudTrail Log:
PatchGroup:
```
{
    "eventVersion": "1.10",
    "userIdentity": {
        "type": "Unknown",
        "accountId": "REDACTED",
        "accessKeyId": "REDACTED"
    },
    "eventTime": "2025-06-18T23:33:51Z",
    "eventSource": "identitystore-scim.amazonaws.com",
    "eventName": "PatchGroup",
    "awsRegion": "eu-west-1",
    "sourceIPAddress": "REDACTED",
    "userAgent": "authentik@2025.6.2",
    "errorCode": "ValidationException",
    "errorMessage": "Current PATCH support allows omitting \"path\" attribute only for Add and Replace operations modifying a single attribute.",
    "requestParameters": {
        "httpBody": {
            "operations": [
                {
                    "op": "REPLACE",
                    "value": "HIDDEN_DUE_TO_SECURITY_REASONS"
                }
            ]
        },
        "tenantId": "REDACTED",
        "id": "REDACTED"
    },
    "responseElements": null,
    "requestID": "REDACTED",
    "eventID": "REDACTED",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "managementEvent": true,
    "recipientAccountId": "REDACTED",
    "eventCategory": "Management",
    "tlsDetails": {
        "clientProvidedHostHeader": "scim.eu-west-1.amazonaws.com"
    }
}
```
PutGroup:
```
{
    "eventVersion": "1.10",
    "userIdentity": {
        "type": "Unknown",
        "accountId": "REDACTED",
        "accessKeyId": "REDACTED"
    },
    "eventTime": "2025-06-18T23:33:51Z",
    "eventSource": "identitystore-scim.amazonaws.com",
    "eventName": "PutGroup",
    "awsRegion": "eu-west-1",
    "sourceIPAddress": "REDACTED",
    "userAgent": "authentik@2025.6.2",
    "errorCode": "NotImplementedException",
    "requestParameters": {
        "tenantId": "REDACTED",
        "id": "REDACTED",
        "requestBody": "HIDDEN_DUE_TO_SECURITY_REASONS"
    },
    "responseElements": null,
    "requestID": "REDACTED",
    "eventID": "REDACTED",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "managementEvent": true,
    "recipientAccountId": "REDACTED",
    "eventCategory": "Management",
    "tlsDetails": {
        "clientProvidedHostHeader": "scim.eu-west-1.amazonaws.com"
    }
}
```
HIDDEN_DUE_TO_SECURITY_REASONS is AWS security thing

@jorhett or @rissson can you please reopen this issue - it's still same response I don't see much reason to open new one for same thing.
Can I help any way? Can I provide you any logs that could help? Config of scim is based on Authentik Docs (+ using AWS Compatibility Mode)

**Comment by Jachimoo** [comment]
@jorhett or @rissson I have the same issue

**Comment by compuguy** [comment]
Hate to add onto this, but I'm running 2025.6.3 and I'm also getting the same issue even with Compatiblity mode set to "AWS". Cloudtrail is still showing it trying to use `PutGroup`, which returns a `NotImplementedException`

**Comment by hkpanchani** [comment]
Same on 2025.6.4

**Comment by FilipK-CZ** [comment]
@rissson Any update? Do you need any more informations?

**Comment by luna-lightblade** [comment]
I did some poking and prodding and the problematic code is in `_update_patch()` which attempts to write the entire group object instead of updating individual attributes. The fallback behavior (which is the default in AWS compatibility mode) also doesn't work at all because AWS IAM Identity Center doesn't support the PUT action for groups (ref: https://docs.aws.amazon.com/singlesignon/latest/developerguide/supported-apis.html)

Given that there are also only 2 attributes other than member which you can even update anyway (ref: https://docs.aws.amazon.com/singlesignon/latest/developerguide/patchgroup.html), I wrote this little monkey patch that sequentially updates those two attributes and then moves on to syncing members.

With this patch, the provider successfully does a full SCIM sync run with the default compatibility mode. 

**THIS PATCH WILL BREAK SYNCING OTHER GROUP ATTRIBUTES WITH NON-AWS SCIM PROVIDERS - YOU HAVE BEEN WARNED.**

```
diff --git a/authentik/providers/scim/clients/groups.py b/authentik/providers/scim/clients/groups.py
index bd387ecdb..617ae3c54 100644
--- a/authentik/providers/scim/clients/groups.py
+++ b/authentik/providers/scim/clients/groups.py
@@ -128,25 +128,29 @@ class SCIMGroupClient(SCIMClient[Group, SCIMProviderGroup, SCIMGroupSchema]):
     def _update_patch(
         self, group: Group, scim_group: SCIMGroupSchema, connection: SCIMProviderGroup
     ):
-        """Update a group via PATCH request"""
-        # Patch group's attributes instead of replacing it and re-adding users if we can
-        self._request(
-            "PATCH",
-            f"/Groups/{connection.scim_id}",
-            json=PatchRequest(
-                Operations=[
-                    PatchOperation(
-                        op=PatchOp.replace,
-                        path=None,
-                        value=scim_group.model_dump(mode="json", exclude_unset=True),
-                    )
-                ]
-            ).model_dump(
-                mode="json",
-                exclude_unset=True,
-                exclude_none=True,
-            ),
-        )
+        """Update a group via PATCH request, only for AWS-allowed attributes (displayName, externalId)."""
+        allowed_attrs = {"displayName", "externalId"}
+        group_dict = scim_group.model_dump(mode="json", exclude_unset=True)
+        for attr in allowed_attrs:
+            if attr in group_dict:
+                self._request(
+                    "PATCH",
+                    f"/Groups/{connection.scim_id}",
+                    json=PatchRequest(
+                        Operations=[
+                            PatchOperation(
+                                op=PatchOp.replace,
+                                path=attr,
+                                value=group_dict[attr],
+                            )
+                        ]
+                    ).model_dump(
+                        mode="json",
+                        exclude_unset=True,
+                        exclude_none=True,
+                    ),
+                )
         return self.patch_compare_users(group)
```

I imagine a more permanent solution would be to replace the current AWS provider compatibility mode with one that applies this behavior rather than the current one that attempts to disable all patching (which is fundamentally flawed anyway since PUT doesn't work on groups).

**Comment by parthapatel0723** [comment]
Is there an update on this? I can't use AWS without having to login with root every time.

**Comment by FilipK-CZ** [comment]
@rissson / @jorhett any updates on this? We're still experiencing the same issues with version 2025.10.1. The documentation lists AWS SCIM Provisioning integration as "Support level: authentik", but it's currently unusable.

**Comment by rissson** [comment]
I'll @BeryJu on this

**Comment by bitpavel-l25** [comment]
Hello everyone. Based on the patch shared by @luna-lightblade I've prepared a PR which uses the correct behaviour only when AWS compatibility mode is enabled.
